Particle samplers and traces for Emcee

philastrophist opened the pull request:
So one of the major gripes I have with pymc3 is that I couldn't use the wonderful emcee from @dfm's emcee sampler very easily with the flexible model specification that pymc3 has. So I've added a generic particle sampler and an emcee sampler along with a trace to be used for particles.

This is somewhat hacky, but it works. There are obvious improvements to be made and a modified emcee sampler could be built in to avoid those pesky reshapes!

However, I have a limited knowledge of the workings of pymc3 despite writing this and so I'm not sure how to incorporate initialisations correctly for the walkers. I want to either use the advi and randomise the result for the walkers, or generate samples from the specified priors (currently doing this).

It'll be great if someone else could look at this!

Thanks

EDIT: Implemented some suggestions and improvements. Still a bit hacky, but I have some ideas

philastrophist committed:
Added generic ParticleStep class and one for Emcee. Also added a backend to handle particle steps. Works but can't display in traceplot

philastrophist committed:
Added generic ParticleStep class and one for Emcee. Also added a backend to handle particle steps. Works but can't display in traceplot

philastrophist committed:
Emcee works with pymc3 now with a little tinkering with start values and a hacky Multitrace.

Still need to flatten shaped variables for emcee to deal with

philastrophist committed:
Emcee works with pymc3 now with a little tinkering with start values and a hacky Multitrace.

Still need to flatten shaped variables for emcee to deal with

philastrophist committed:
added shape handling to emcee sampler, needed to swap first two storage dimensions

philastrophist committed:
turned off storechain in emcee

philastrophist commented:
An example:

```
import pymc3 as pm
import numpy as np
import matplotlib.pyplot as plt
from pymc3.external.emcee_samplers import EmceeEnsemble

x = np.linspace(0, 1, 100)
y = 3 + (x*2)
y += np.random.normal(0, 0.5, size=len(x))
yerr = np.ones_like(y) * 0.1

with pm.Model() as model:
    eps = pm.HalfCauchy('eps', beta=10)

    intercept = pm.Normal('intercept', mu=0, sd=10)
    gradient = pm.Normal('gradient', mu=0, sd=10)
    obs = pm.Normal('obs', mu=y, sd=yerr, shape=len(y))
    line = pm.Deterministic('line', intercept + (x * gradient))
    like = pm.Normal('like', mu=line, sd=eps, observed=obs)
    chain = pm.sample(5000, init='advi', step=EmceeEnsemble())

i = np.percentile(chain['intercept'], [16, 50, 84])
g = np.percentile(chain['gradient'], [16, 50, 84])
e = np.percentile(chain['eps'], [16, 50, 84])

print i
print g
print e

pm.traceplot(chain[::10], varnames=['eps', 'intercept', 'gradient'])
```

philastrophist committed:
add emcee to test environment

philastrophist committed:
used to fail for non-compound step methods

philastrophist committed:
used to fail for non-compound step methods

philastrophist committed:
removed baggage

philastrophist commented:
There was a previous proposal about this: https://github.com/pymc-devs/pymc3/issues/659

fonnesbeck commented on pymc3/sampling.py:
```diff
@@ -83,7 +85,7 @@ def assign_step_methods(model, step=None, methods=(NUTS, HamiltonianMC, Metropol
 def sample(draws, step=None, init='advi', n_init=200000, start=None,
            trace=None, chain=0, njobs=1, tune=None, progressbar=True,
-           model=None, random_seed=-1):
+           model=None, random_seed=-1, ):
```
Trailing comma

fonnesbeck commented on pymc3/step_methods/particle.py:
```diff
@@ -0,0 +1,73 @@
+        return bij.rmap(apoint)
+
+
+class EmceeSamplerStep(ParticleStep):
```
Scratch the last comment (deleted), it is using emcee directly. In that case, I think this should be in `pymc3.external`, just as the Edward extension is.

fonnesbeck replied:
Our convention is not to include `Step` in the name of user-facing sampler classes, so I suggest renaming to simply `Emcee`.

fonnesbeck commented:
Thanks for the PR; I had a quick peek at the code and will look at it in depth later. I'm delighted to see the addition of particle samplers, in general. As I commented above, my own preference would be to put the `EmceeSampler` class in the `external` submodule, which is where we keep the Edward extension, and have emcee as a soft dependency.

ferrine commented on pymc3/step_methods/particle.py:
```diff
@@ -0,0 +1,73 @@
+
+    f = theano.function([inarray0], logp0)
+    f.trust_input = True
+    return f
```
new line

philastrophist commented:
I'll get onto those suggestions listed above, but I'm thinking now that there could be a factory function to build a step around an external sampler. This would remove the dependency on emcee and allow easy addition of other samplers by user, no?

philastrophist committed:
aesthetics

philastrophist committed:
aesthetics

twiecki commented:
@philastrophist Very cool. Did you look at https://github.com/pymc-devs/pymc3/pull/1569?

springcoil commented:
Very cool stuff @philastrophist. A PhD student I mentored at a previous job was very upset that particle samplers weren't in PyMC3 - I'll let him know.

philastrophist committed:
added `.random` for transformed distributions

philastrophist committed:
added defaults for emcee sampler

philastrophist committed:
added defaults for emcee sampler

philastrophist committed:
moved `Emcee` to externals. But left ParticleStep

philastrophist committed:
Explicit NotImplementedException for mixtures of ParticleStep and other steps. Can only be used by itself at the moment!

twiecki commented:
@philastrophist also needs tests and an example NB for the docs.

philastrophist committed:
edited init_nuts to partition into do_init and init_nuts. This allows us to initialise ParticleSteps really simply in the sample function. All aspects remain the same expect that a randomiser is now included to disperse the initial start values for the particles. This can also be used for normal samplers and defaults to the original method.
`expected_ndraws` included in the step to make that bit easier for emcee. Could figure a better way out later.

philastrophist committed:
edited init_nuts to partition into do_init and init_nuts. This allows us to initialise ParticleSteps really simply in the sample function. All aspects remain the same expect that a randomiser is now included to disperse the initial start values for the particles. This can also be used for normal samplers and defaults to the original method.
`expected_ndraws` included in the step to make that bit easier for emcee. Could figure a better way out later.

philastrophist committed:
added point to multindarray

philastrophist committed:
added point to multindarray

philastrophist committed:
added some todos and doc string explanations in prep for some proper documentation

philastrophist committed:
removed some silly ways to give emcee `iterations` information and tidied up

philastrophist committed:
removed some silly ways to give emcee `iterations` information and tidied up

philastrophist commented:
Now the particle samplers can use the `advi`/`map` etc easily in the call to `pm.sample`! 

Not read https://github.com/pymc-devs/pymc3/issues/1569 yet but plan to see how it works.

Tests and the NB are underway. For now I'll update the example above.

There is currently somewhat more lines than strictly necessary in `sample` but I'll prune those in the near future. I currently have it so  that you can have multiple multiprocessing jobs along with multiple particles. I don't know why you would want to do that but the functionality is there. I might remove it though to clean up some confusing sections.
I'm almost there with integrating other samplers into a `ParticleCompoundStep` for those scenarios that require something like discrete choice (emcee doesn't do that).

All in all it's basically done and works. What I'm currently doing is making sure it can do all the things current samplers can whilst incorporating some flexibility for people to add other `ParticleStep`s.

P.S. This started out as something I did for myself and I'm doing it when I have time. So it may take time for me to finish everything!

philastrophist committed:
corrected overwrite of a given `start` value

philastrophist committed:
corrected overwrite of a given `start` value

twiecki commented:
@philastrophist Seems like you need to rebase.

fonnesbeck commented:
Great, just remove the WIP from the tag and title when you want a pre-merge review.

philastrophist committed:
stopped overwrite of given trace

philastrophist committed:
allow random init to be reached

philastrophist committed:
start particle shape handling and test

philastrophist committed:
added more tests

philastrophist committed:
added init_start

philastrophist committed:
simplified transform_start

philastrophist committed:
fixed init and start

philastrophist committed:
forgot a `not`

philastrophist committed:
consistency between backends and multisqlite

philastrophist committed:
all steps have default `nparticles = None`

twiecki commented:
@philastrophist Any progress?

philastrophist commented:
Currently the way this method works is quite restrictive on how the sampler works. So it only works for 1 particle sampler at a time. I think this is fine for now since any further development will need the multindtrace. 
I need to rebase but i'll push the latest version very soon. 
I've made some progress on generalising particle samplers in a separate branch which I'd like some feedback on once I've finished it.

philastrophist committed:
fix transform_particles

philastrophist committed:
init available to all step methods now

philastrophist committed:
added notebook

philastrophist committed:
todo

philastrophist committed:
fixed init=None problem

philastrophist committed:
added documentation notebook

philastrophist committed:
fixed issue with starting points for deterministic variables

philastrophist commented:
I have made some subtle `pm.sample` api changes with regard to initialisation and start points to make it more general (and so it can be applied to `particlestep`s easily). However, this means it breaks in a small subset of cases (4 tests fail) where start is specified. The error thrown relates to variable type.

Since this is a minor problem with a change in api, I think now is the time to remove WIP (since it works for any of my cases) and start some discussion...

PS. A notebook for an intro into how to use `EmceeEnsemble` is included in the last commit

twiecki commented on pymc3/sampling.py:
```diff
@@ -453,41 +596,52 @@ def init_nuts(init='ADVI', njobs=1, n_init=500000, model=None,
-    return start, step
+def transform_start_particles(start, nparticles, model=None):
+    """
+    :param start:
```
These doc-strings need to be made numpy-style.

philastrophist replied:
That's my editor's default. Will change

twiecki commented on pymc3/sampling.py:
```diff
@@ -453,41 +596,52 @@ def init_nuts(init='ADVI', njobs=1, n_init=500000, model=None,
-            start = start[0]
-    else:
-        raise NotImplemented('Initializer {} is not supported.'.format(init))
+def get_random_starters(nwalkers, model=None):
```
Can this only be used with a particle sampler? Seems like the function is more general than that.

philastrophist replied:
yes, I'll move it outside the particlestep block

twiecki commented on pymc3/sampling.py:
```diff
@@ -453,41 +596,52 @@ def init_nuts(init='ADVI', njobs=1, n_init=500000, model=None,
-    step = pm.NUTS(scaling=cov, is_cov=True, **kwargs)
 
-    return start, step
+def transform_start_particles(start, nparticles, model=None):
```
should this be moved to external/emcee?

philastrophist replied:
I feel like this should stay here since it's applicable to specifying jobs and other particle steps not just emcee

twiecki commented:
I think this looks good overall. I'm a bit concerned of the added code complexity but the particle stuff should be usable by other samples except emcee too.

The changes to the API make sense.

It does need tests however.

philastrophist commented:
Thanks for the feedback. I've made some comments, but what in particular are you concerned about with added code complexity? The additions to `.sample`? Or the duplication of `NDArray`? 

With respect to `NDArray`, I could merge `MultiNDArray` and `NDarray` by making particles=1 the default instead of having two separate classes for `particles=None` and `particles!=None`. 
I guess since I have implemented `nparticles` in all steps it would make sense to have nparticles=1 as the default and merge the different trace styles.

Tests will be implemented of course, I just wanted to get the api out of the way first.

philastrophist committed:
simplified sample
allowed other steps to use init='random'

philastrophist committed:
fixed issue with nsamples=1 giving 2d start variables

philastrophist committed:
simplified sample again

twiecki commented on pymc3/backends/sqlite.py:
```diff
@@ -260,6 +260,122 @@ def point(self, idx):
+        ----------
+        varname : str
+        burn : int
+        thin : int
```
missing particles doc

twiecki commented on pymc3/backends/ndarray.py:
```diff
@@ -177,7 +180,123 @@ def point(self, idx):
 
 
-def _slice_as_ndarray(strace, idx):
+class MultiNDArray(NDArray, base.MultiMixin):
```
Needs doc-string.

philastrophist committed:
moved definition of nparticles into pm.sample from sampler __init__

philastrophist committed:
moved min_nparticles and nparticles to BlockedStep definition

philastrophist committed:
docstrings

philastrophist committed:
moved transform_start_particles

philastrophist committed:
docstring for emcee

philastrophist committed:
fixed assertion message

philastrophist committed:
adapted test. all backends would be called with nparticles=>something<

philastrophist committed:
fixed wrong __len__ specification

philastrophist committed:
pm.sample is inside model context now. Changed test init to adapt to change in pm.sample(init='advi')

philastrophist committed:
removed confusing stuff

philastrophist commented:
There is a test: `test_step.TestStepMethods.test_sample_exact(NUTS)` that fails specifically for `NUTS`. 
The arrays don't match where x is the master "known sample" and y is the one generated.

```
x: array([  1.203436e+00,   1.203436e+00,   3.284390e-01,  -5.632208e-01,
         1.374194e+00,   1.729373e+00,   4.398738e-01,   4.398738e-01,
        -4.584570e-01,  -3.028040e-01,   9.709908e-01,   1.390899e+00,...
 y: array([ 1.118324,  1.118324,  0.428752, -1.063373, -1.175014, -0.277752,
       -0.277752,  0.003286,  0.23796 ,  1.655296,  1.790141, -0.623905,
        0.60124 ,  0.60124 ,  0.684407,  0.704353,  0.602374,  0.037173,...
```

I'm not quite sure what to make of this. 

Other than that I think I have, or can easily, fix the errors raised in testing for this branch. Now I'll just put in a `MultiNDArray` test and we'll be set.

philastrophist committed:
alterations to make tests pass. working on better init default value

philastrophist committed:
better default init value (auto)

twiecki commented:
@philastrophist Any progress?

philastrophist commented:
Still going... Rectifying errors

philastrophist committed:
refactored to separate normal steps and particle step.
`transform_start_positions` now accounts for njobs

philastrophist commented:
Tests just underway

philastrophist committed:
init must be None for non-continuous

philastrophist committed:
brackets

reac2 commented:
Just wondering - what's the status with the Emcee sampler? Excited to test this out on our data.

twiecki commented:
Good question, would need to ask @philastrophist.

@reac2 is your model non-differentiable?

reac2 commented:
Thanks - will do!
Sadly yes and currently seems to look like it has converged after relatively lengthly chains but away from some of the correct values - was hoping the many walker approach would explore parameter space more efficiently.  Will try implementing something similar to your 2013 hack too http://twiecki.github.io/blog/2013/09/23/emcee-pymc/

twiecki commented:
Have you tried the new ATMCMC sampler?

twiecki commented:
https://github.com/pymc-devs/pymc3/blob/7fb2b55bdd1a11123b541e5b1a7825b864dacc79/pymc3/step_methods/smc.py#L76

reac2 commented:
Ah, no I was looking for ATMCMC and thought it had been removed. That's great - I will try this one out! Thanks very much.

reac2 commented:
Is there an example for the ATMCMC code you can point me to? Our model is similar in form to the arbitrary deterministic disasters example so i'm starting there but running into trouble with redefining the likelihood_name parameter in smc.SMC.

ColCarroll commented:
I'd check out the test case in `test_smc.py`.  There's another super basic one in `test_step.py`.  I'd be very interested for more examples of using it, and how to make the API better.  Note -- still working on checking it thoroughly for accuracy, too!

reac2 commented:
Great, found it! Thanks. I'll see if I can get this going with the arbitrary deterministic disasters model.

junpenglao commented:
@reac2 you can also find an example here fitting a mixed-effect model https://github.com/junpenglao/GLMM-in-Python/blob/master/pymc3_different_sampling.py#L86-L113

philastrophist commented:
Please see https://github.com/pymc-devs/pymc3/pull/2253 for the up-to-date refactored version. I consider this pull request outdated and will close it.
